lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,918 @@
+			//  - MUST fail the negotiation if:
+			//     - if has received 4096 `tx_add_output` messages during this negotiation
+			return Err(AbortReason::ReceivedTooManyTxAddOutputs);
+		}
+
+		if msg.sats < msg.script.dust_value().to_sat() {
+			// The receiving node:
+			// - MUST fail the negotiation if:
+			//		- the sats amount is less than the dust_limit
+			return Err(AbortReason::BelowDustLimit);
+		}
+		if msg.sats > TOTAL_BITCOIN_SUPPLY_SATOSHIS {
+			// The receiving node:
+			// - MUST fail the negotiation if:
+			//		- the sats amount is greater than 2,100,000,000,000,000 (TOTAL_BITCOIN_SUPPLY_SATOSHIS)
+			return Err(AbortReason::ExceededMaximumSatsAllowed);
+		}
+
+		// The receiving node:
+		//   - MUST accept P2WSH, P2WPKH, P2TR scripts
+		//   - MAY fail the negotiation if script is non-standard
+		if !msg.script.is_v0_p2wpkh() && !msg.script.is_v0_p2wsh() && !msg.script.is_v1_p2tr() {
+			return Err(AbortReason::InvalidOutputScript);
+		}
+
+		if self.outputs.iter().any(|(serial_id, _)| *serial_id == msg.serial_id) {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//    - the `serial_id` is already included in the transaction
+			return Err(AbortReason::DuplicateSerialId);
+		}
+
+		let output = TxOut {
+			value: msg.sats,
+			script_pubkey: msg.script.clone(),
+		};
+		self.outputs.insert(msg.serial_id, output);
+		Ok(())
+	}
+
+	fn remote_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) -> Result<(), AbortReason> {
+		if !self.is_serial_id_valid_for_counterparty(&msg.serial_id) {
+			return Err(AbortReason::IncorrectSerialIdParity);
+		}
+		if let Some(_) = self.outputs.remove(&msg.serial_id) {
+			Ok(())
+		} else {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//    - the input or output identified by the `serial_id` was not added by the sender
+			//    - the `serial_id` does not correspond to a currently added input
+			Err(AbortReason::SerialIdUnknown)
+		}
+	}
+
+	fn local_tx_add_input(&mut self, msg: &msgs::TxAddInput) {
+		let tx = msg.prevtx.clone().into_transaction();
+		let input = TxIn {
+			previous_output: OutPoint {
+				txid: tx.txid(),
+				vout: msg.prevtx_out,
+			},
+			sequence: Sequence(msg.sequence),
+			..Default::default()
+		};
+		debug_assert!((msg.prevtx_out as usize) < tx.output.len());
+		let prev_output = &tx.output[msg.prevtx_out as usize];
+		self.prevtx_outpoints.insert(input.previous_output.clone());
+		self.inputs.insert(msg.serial_id, TxInputWithPrevOutput {
+			input,
+			prev_output: prev_output.clone(),
+		});
+	}
+
+	fn local_tx_add_output(&mut self, msg: &msgs::TxAddOutput) {
+		self.outputs.insert(msg.serial_id, TxOut {
+			value: msg.sats,
+			script_pubkey: msg.script.clone(),
+		});
+	}
+
+	fn local_tx_remove_input(&mut self, msg: &msgs::TxRemoveInput) {
+		self.inputs.remove(&msg.serial_id);
+	}
+
+	fn local_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) {
+		self.outputs.remove(&msg.serial_id);
+	}
```

Comment:
Do we ever check that user-provided inputs and outputs are valid?

Reply:
@dunxen

Reply:
We do some checks on the dual funding branch through the API where the user contributes those inputs. So we don't do anything here. Either the same method or a similar API would be used for splicing extra user inputs/outputs.